lib/AST/NameLookup.cpp

```diff
@@ -1087,7 +1087,8 @@ static DirectlyReferencedTypeDecls
 directReferencesForTypeRepr(Evaluator &evaluator, ASTContext &ctx,
                             TypeRepr *typeRepr, DeclContext *dc,
                             bool allowUsableFromInline,
-                            bool rhsOfSelfRequirement);
+                            bool rhsOfSelfRequirement,
+                            bool allowProtocolMembers);
```

Comment:
It’s probably time to replace these boolean flags with an OptionSet.

Reply:
@slavapestov There are a couple of reasons I wanted to avoid doing that with this PR:

1. In my opinion, this kind of cycle should be avoided in a more structured way by ASTScope lookup. It is obvious that if it walks up to a parent scope and enumerates the children, it will find the scope it just visited, and depending on which methods happen to be called, it is incredibly easy for it to get in to a request cycle. If it knew the child it had just visited, it would be able to avoid this.

    I tried to implement this, but there's quite a bit of plumbing to do, and of course I don't know if there are any longer-term plans to rework this lookup so I wasn't sure if it was worth doing right now. I'd like to keep this change low-risk so it can be cherry-picked to 6.0, and major refactoring of name lookup seems to me higher risk than is absolutely necessary to fix this issue.

2. I haven't audited all of the places which pass `allowProtocolMembers` as true. Right now, everywhere passes it as `true` in order to maintain the behaviour before this patch, but I suspect some places could afford to turn it off for a slight performance improvement.

Also, could you run the smoke test again? There was a failure on Mac, but the data isn't available on CI any more (I think it was something to do with LLDB? Possibly unrelated to this change. I tried the LLDB tests locally and don't see anything related)

Reply:
Ping @slavapestov